libbeat/common/datetime.go

```diff
@@ -23,45 +23,72 @@ import (
 	"errors"
 	"hash"
 	"time"
+
+	"github.com/elastic/beats/v7/libbeat/common/dtfmt"
 )
 
-// TsLayout is the layout to be used in the timestamp marshaling/unmarshaling everywhere.
-// The timezone must always be UTC.
-const TsLayout = "2006-01-02T15:04:05.000Z"
+const (
+	// TsLayout is the seconds layout to be used in the timestamp marshaling/unmarshaling everywhere.
+	// The timezone must always be UTC.
+	TsLayout = "2006-01-02T15:04:05.000Z"
+
+	tsLayoutMillis = "2006-01-02T15:04:05.000Z"
+	tsLayoutMicros = "2006-01-02T15:04:05.000000Z"
+	tsLayoutNanos  = "2006-01-02T15:04:05.000000000Z"
+)
 
 // Time is an abstraction for the time.Time type
 type Time time.Time
 
+var defaultTimeFormatter = dtfmt.MustNewFormatter("yyyy-MM-dd'T'HH:mm:ss.fffffffff'Z'")
+
+var defaultParseFormats = []string{
+	tsLayoutMillis,
+	tsLayoutMicros,
+	tsLayoutNanos,
+}
+
 // MarshalJSON implements json.Marshaler interface.
 // The time is a quoted string in the JsTsLayout format.
 func (t Time) MarshalJSON() ([]byte, error) {
-	return json.Marshal(time.Time(t).UTC().Format(TsLayout))
+	str, _ := defaultTimeFormatter.Format(time.Time(t).UTC())
+	return json.Marshal(str)
 }
 
 // UnmarshalJSON implements js.Unmarshaler interface.
 // The time is expected to be a quoted string in TsLayout
 // format.
 func (t *Time) UnmarshalJSON(data []byte) (err error) {
 	if data[0] != []byte(`"`)[0] || data[len(data)-1] != []byte(`"`)[0] {
-		return errors.New("Not quoted")
+		return errors.New("not quoted")
 	}
 	*t, err = ParseTime(string(data[1 : len(data)-1]))
-	return
+	return err
 }
 
 func (t Time) Hash32(h hash.Hash32) error {
 	err := binary.Write(h, binary.LittleEndian, time.Time(t).UnixNano())
 	return err
 }
 
-// ParseTime parses a time in the TsLayout format.
+// ParseTime parses a time in the NanoTsLayout format first, then use millisTsLayout format
 func ParseTime(timespec string) (Time, error) {
-	t, err := time.Parse(TsLayout, timespec)
+	var err error
+	var t time.Time
+
+	for _, layout := range defaultParseFormats {
```

Comment:
It looks like this would try Millis, then Micros, then finally Nanos unless I'm missing something from above:

```go
var defaultParseFormats = []string{
	tsLayoutMillis,
	tsLayoutMicros,
	tsLayoutNanos,
}
```

Reply:
Also, do we expect any performance implications from possibly parsing the time up to 3x?

Reply:
When running the Beat, no. `ParseTime` is only used through `MustParseTime` in unit tests. Those already passed without panic, so the times passed to `ParseTime` could be parsed on the first try. So the change will not have an effect on existing tests. Only those test will go through the 3 hops to parse the time that contain nanosecond precitions timestamps.